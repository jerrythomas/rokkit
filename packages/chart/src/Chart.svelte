<script>
	import { run } from 'svelte/legacy';

	import { setContext } from 'svelte'
	import { writable } from 'svelte/store'
	import { chart } from './lib'

	let config = writable({})
	setContext('chart', config)

	
	/**
	 * @typedef {Object} Props
	 * @property {any} data
	 * @property {any} x
	 * @property {any} y
	 * @property {any} [value]
	 * @property {any} [color]
	 * @property {any} [fill]
	 * @property {number} [width] - export let pattern = x
	 * @property {number} [height]
	 * @property {any} [padding]
	 * @property {number} [marginLeft]
	 * @property {number} [marginRight]
	 * @property {number} [marginTop]
	 * @property {number} [marginBottom]
	 * @property {boolean} [flipCoords]
	 * @property {number} [spacing]
	 * @property {import('svelte').Snippet} [children]
	 */

	/** @type {Props} */
	let {
		data,
		x,
		y,
		value = y,
		color = x,
		fill = x,
		width = 2048,
		height = 2048,
		padding = height / 16,
		marginLeft = 0,
		marginRight = 0,
		marginTop = 0,
		marginBottom = 0,
		flipCoords = false,
		spacing = 0.1,
		children
	} = $props();

	let margin = $derived({
		left: marginLeft,
		right: marginRight,
		top: marginTop,
		bottom: marginBottom
	})
	// $: patterns = [...new Set(data.map((d) => d[pattern]))]
	// $: fills = [...new Set(data.map((d) => d[fill]))]
	// $: colors = [...new Set(data.map((d) => d[color]))]
	run(() => {
		config.set(
			chart(data, {
				x,
				y,
				value,
				color,
				fill,
				width,
				height,
				padding,
				margin,
				flipCoords,
				spacing
			})
		)
	});
</script>

<svg
	viewBox="0 0 {$config.width} {$config.height}"
	width={$config.width}
	height={$config.height}
	class="w-full h-full chart"
>
	<rect
		x="0"
		y="0"
		width={$config.width}
		height={$config.height}
		fill="none"
		stroke="currentColor"
	/>
	{@render children?.()}
</svg>
