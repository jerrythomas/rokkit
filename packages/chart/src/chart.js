// export function scales(data, aes, opts) {
// 	const { x, y } = aes
// 	const { width, height, flipCoords } = {
// 		width: 800,
// 		height: 600,
// 		flipCoords: false,
// 		...opts
// 	}

// 	return { x, y, width, height, flipCoords }
// }
