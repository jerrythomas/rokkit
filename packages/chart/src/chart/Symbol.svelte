<script>
	import { namedShapes } from '../components/lib/shape'

	export let x = 0
	export let y = 0
	export let size = 10
	export let fill = 'none'
	export let stroke = 'currentColor'
	export let thickness = 0.5

	export let name = null
	export let shape = null

	$: x = x - size / 2
	$: y = y - size / 2

	$: d =
		typeof shape === 'function'
			? shape(size)
			: (shape || name) in namedShapes
			? namedShapes[shape || name](size)
			: namedShapes.circle(size)
</script>

<!-- svelte-ignore a11y-click-events-have-key-events -->
<path
	{d}
	{fill}
	{stroke}
	transform="translate({x},{y})"
	stroke-width={thickness}
	fill-rule="evenodd"
	on:click
	on:mouseover
	on:mouseleave
	on:focus
/>
