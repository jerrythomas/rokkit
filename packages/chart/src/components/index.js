export { toHexString, initCap, uniqueId, toNested } from './lib/utils'
export { brewer, uniques, slidingWindow, colors } from './lib/index'
