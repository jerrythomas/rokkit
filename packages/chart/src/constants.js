export const __muted__ = {
	color: '#eeeeee',
	fill: 'empty',
	shape: 'circle'
}
export const __colors__ = [
	'#FFDE6B',
	'#EF89EE',
	'#F79F1E',
	'#02B8FF',
	'#9F84EC',
	'#15CBC4',
	'#0092FD',
	'#F63A57',
	'#A2CB39',
	'#FF6E2F',
	'#FEB8B9',
	'#af7aa1',
	'#7EFFF5'
]

export const __patterns__ = {
	A: 'M0 5A6 6 0 0 0 10 5',
	B: 'M0 10L10 0',
	C: 'M0 0A10 10 0 0 0 10 10',
	D: 'M0 0L10 10',
	E: 'M10 5A6 6 0 0 0 0 5',
	F: 'M10 10A10 10 0 0 0 0 0',
	G: 'M0 0L10 10ZM10 0L0 10Z',
	H: 'M1 1L9 1L9 9L1 9Z',
	I: 'M4 0L4 10M6 10L6 0M0 4L10 4M10 6L0 6',
	J: 'M0 2L8 10M2 0L10 8M0 8L8 0M2 10L10 2',
	K: 'M5 1A 4 4 0 0 0 9 5A4 4 0 0 0 5 9A4 4 0 0 0 1 5A4 4 0 0 0 5 1',
	L: 'M1 3L7 9M3 1L9 7M1 7L7 1M3 9L9 3',
	M: 'M2 2A4 4 0 0 0 8 2A4 4 0 0 0 8 8A4 4 0 0 0 2 8A4 4 0 0 0 2 2',
	N: 'M0 0A5 5 0 0 0 10 0A5 5 0 0 0 10 10A5 5 0 0 0 0 10A5 5 0 0 0 0 0',
	O: 'M5 2A 3 3 0 0 0 8 5A3 3 0 0 0 5 8A3 3 0 0 0 2 5A3 3 0 0 0 5 2',
	P: 'M2 5L5 2L8 5L5 8Z',
	Q: 'M3 5A2 2 0 0 0 7 5A2 2 0 0 0 3 5M1 5L9 5M5 1L5 9',
	R: 'M2 8L8 2ZM1.5 3.5L3.5 1.5ZM6.5 8.5L8.5 6.5ZM0 0L10 10Z',
	S:
		'M2 8L8 2ZM1.5 3.5L3.5 1.5Z' +
		'M6.5 8.5L8.5 6.5Z' +
		'M2 2L8 8M1.5 6.5L3.5 8.5' +
		'M6.5 1.5L8.5 3.5',
	T:
		'M5 1 A6 6 0 0 0 5 9' +
		'A6 6 0 0 0 5 1' +
		'M1 5A6 6 0 0 0 9 5A6 6 0 0 0 1 5',
	U:
		'M1.5 5A1 1 0 0 0 3.5 5A1 1 0 0 0 1.5 5' +
		'M6.5 5A1 1 0 0 0 8.5 5A1 1 0 0 0 6.5 5' +
		'M5 1.5A1 1 0 0 0 5 3.5A1 1 0 0 0 5 1.5' +
		'M5 6.5A1 1 0 0 0 5 8.5A1 1 0 0 0 5 6.5',
	V:
		'M1.5 2.5A1 1 0 0 0 3.5 2.5A1 1 0 0 0 1.5 2.5' +
		'M6.5 2.5A1 1 0 0 0 8.5 2.5A1 1 0 0 0 6.5 2.5' +
		'M2.5 6.5A1 1 0 0 0 2.5 8.5A1 1 0 0 0 2.5 6.5' +
		'M7.5 6.5A1 1 0 0 0 7.5 8.5A1 1 0 0 0 7.5 6.5' +
		'M3.5 5A1 1 0 0 0 6.5 5A1 1 0 0 0 3.5 5',
	W: 'M5 0L6 4L10 5L6 6L5 10L4 6L0 5L4 4Z' + 'M2 1V3M1 2H3' + 'M8 9V7M9 8H7',
	X: 'M5 2L2.5 9L8.8 4.6L1.2 4.6L7.5 9Z',
	Y: 'M0 5A5 5 0 0 0 5 0' + 'M5 10A5 5 0 0 0 0 5' + 'M5 10A5 5 0 0 0 5 0',
	Z: 'M0 0L10 10M5 0L10 5M0 5 L5 10',
	Z1: 'M0 0L10 10M3 0L10 7M0 7 L3 10'
}
