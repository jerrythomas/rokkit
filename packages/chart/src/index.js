// export { aes } from './lib/aes'
// export { timelapse } from './lib/timelapse'
// export { axisTicks } from './lib/axis'
export { timer, elapsed } from './lib/timer'
// export { default as Chart } from './chart/Chart.svelte'
// export { default as Axis } from './chart/Axis.svelte'
export { default as Swatch } from './chart/Swatch.svelte'
// export { default as AxisTicks } from './chart/AxisTicks.svelte'
// export { default as AxisGrid } from './chart/AxisGrid.svelte'
// export { default as AxisLabels } from './chart/AxisLabels.svelte'
export { default as BarPlot } from './plots/BarPlot.svelte'
export { default as LinePlot } from './plots/LinePlot.svelte'
export { default as BoxPlot } from './plots/BoxPlot.svelte'
export { default as ViolinPlot } from './plots/ViolinPlot.svelte'
export { default as ScatterPlot } from './plots/old_ScatterPlot.svelte'
export * from './components/lib'
