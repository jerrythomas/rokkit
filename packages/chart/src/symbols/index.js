import { namedShapes } from './constants'
import { default as Shape } from './Shape.svelte'
import { default as RoundedSquare } from './RoundedSquare.svelte'

export const shapes = [...Object.keys(namedShapes), 'rounded-square']
export const components = {
	default: Shape,
	'rounded-square': RoundedSquare
}
