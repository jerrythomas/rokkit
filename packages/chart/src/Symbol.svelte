<script>
	import { components } from './symbols'

	export let x = 0
	export let y = 0
	export let size = 10
	export let fill = 'currentColor'
	export let stroke = 'currentColor'
	export let shape = 'circle'
	/** @type Object<string, any> */
	export let using = components

	$: component = using[shape] || using.default
	$: props = using[shape] ? $$restProps : { name: shape, ...$$restProps }
</script>

<svelte:component this={component} {x} {y} {size} {fill} {stroke} {...props} />
