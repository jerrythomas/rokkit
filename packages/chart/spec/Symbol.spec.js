import { describe, expect, beforeEach, it } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import Symbol from '../src/Symbol.svelte'
import { Triangle } from '../src/symbols'

describe('Symbol.svelte', () => {
	beforeEach(() => cleanup())

	it('should render default circle.', () => {
		const { container } = render(Symbol)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render symbol provided', () => {
		const { container } = render(Symbol, {
			shape: 'Triangle',
			using: { Triangle }
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render using path when component is not provided', () => {
		const { container } = render(Symbol, { shape: 'triangle', using: {} })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
})
