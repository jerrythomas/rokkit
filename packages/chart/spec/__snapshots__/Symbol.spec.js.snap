// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Symbol.svelte > should render default circle. 1`] = `
<body>
  <div>
    <circle
      cx="0"
      cy="0"
      fill="currentColor"
      r="5"
      role="button"
      stroke="currentColor"
      tabindex="0"
    />
    
  </div>
</body>
`;

exports[`Symbol.svelte > should render symbol provided 1`] = `
<body>
  <div>
    
  </div>
</body>
`;

exports[`Symbol.svelte > should render using path when component is not provided 1`] = `
<body>
  <div>
    
  </div>
</body>
`;
